File: solutions/ruby/meetup/1/meetup.rb
require 'date'

class Meetup
  DAYS = %i[ sunday monday tuesday wednesday thursday friday saturday ].freeze

  DAYS_PER_WEEK = 7
  WEEK_NUMBERS  = %i[ first second third fourth ].freeze
  TEENTHS       = [ 13, 14, 15, 16, 17, 18, 19 ]

  def initialize(month, year)
    @month = month
    @year  = year
  end

  def day(day_of_week, week)
    first_of_month = Date.new(@year, @month, 1)

    offset = DAYS.index(day_of_week) - first_of_month.wday
    offset += DAYS_PER_WEEK if offset.negative?

    first_day = first_of_month.next_day(offset)

    return last(first_day)   if week == :last
    return teenth(first_day) if week == :teenth

    first_day.next_day(DAYS_PER_WEEK * WEEK_NUMBERS.index(week))
  end

  def last(current_day)
    return current_day.prev_day(DAYS_PER_WEEK) if current_day.month != @month
    last(current_day.next_day(DAYS_PER_WEEK))
  end

  def teenth(current_day)
    return current_day if TEENTHS.include?(current_day.day)
    teenth(current_day.next_day(DAYS_PER_WEEK))
  end
end

File: solutions/javascript/meetup/1/meetup.js
const daysPerWeek = 7;
const weekPlaces  = [ 'first', 'second', 'third', 'fourth' ];
const teenths     = [ 13, 14, 15, 16, 17, 18, 19 ];

const getDayIndex = (day) => {
  switch (day) {
    case 'Sunday':    return 0;
    case 'Monday':    return 1;
    case 'Tuesday':   return 2;
    case 'Wednesday': return 3;
    case 'Thursday':  return 4;
    case 'Friday':    return 5;
    case 'Saturday':  return 6;
    default: throw new Error('Invalid day.  Be sure to use proper name casing.');
  }
}

const getDayOffset = (day, index) => {
  const offset = getDayIndex(day) - index;
  return offset < 0 ? offset + daysPerWeek : offset;
}

const getLastDay = (day, monthIndex) => {
  if (day.getMonth() !== monthIndex) {
    day.setDate(day.getDate() - daysPerWeek);
    return day;
  }

  day.setDate(day.getDate() + daysPerWeek);
  return getLastDay(day, monthIndex);
}

const getTeenthDay = (day) => {
  if (teenths.includes(day.getDate())) {
    return day;
  }

  day.setDate(day.getDate() + daysPerWeek);
  return getTeenthDay(day);
}

export const meetup = (year, month, week, day) => {
  const monthIndex = month - 1;
  const firstOfMonth = new Date(year, monthIndex, 1);
  const offset = getDayOffset(day, firstOfMonth.getDay());

  let currentDay = firstOfMonth;
  currentDay.setDate(currentDay.getDate() + offset);

  if (week === 'last')   return getLastDay(currentDay, monthIndex);
  if (week === 'teenth') return getTeenthDay(currentDay);

  currentDay.setDate(
    currentDay.getDate() + daysPerWeek * weekPlaces.indexOf(week)
  );

  return currentDay;
};
